x/smart/stream_dialer.go

```diff
@@ -292,23 +309,60 @@ func (f *StrategyFinder) findTLS(ctx context.Context, testDomains []string, base
-// It returns an error if no strategy was found that unblocks the testDomains.
-// The testDomains must be domains with a TLS service running on port 443.
-func (f *StrategyFinder) NewDialer(ctx context.Context, testDomains []string, configBytes []byte) (transport.StreamDialer, error) {
-	var parsedConfig configConfig
-	err := yaml.Unmarshal(configBytes, &parsedConfig)
-	if err != nil {
-		return nil, fmt.Errorf("failed to parse config: %v", err)
+// Return the fastest fallback dialer that is able to access all the testDomans
+func (f *StrategyFinder) findFallback(ctx context.Context, testDomains []string, fallbackConfig []string) (transport.StreamDialer, error) {
+	if len(fallbackConfig) == 0 {
+		return nil, errors.New("no fallback was specified")
 	}
 
-	// Make domain fully-qualified to prevent confusing domain search.
-	testDomains = append(make([]string, 0, len(testDomains)), testDomains...)
-	for di, domain := range testDomains {
-		testDomains[di] = makeFullyQualified(domain)
+	ctx, searchDone := context.WithCancel(ctx)
+	defer searchDone()
+	raceStart := time.Now()
+	type SearchResult struct {
+		Dialer transport.StreamDialer
+		Config string
+	}
+
+	fallback, err := raceTests(ctx, 250*time.Millisecond, fallbackConfig, func(fallbackUrl string) (*SearchResult, error) {
+		dialer, err := getStreamDialer(ctx, fallbackUrl)
+		if err != nil {
+			return nil, fmt.Errorf("getStreamDialer failed: %w", err)
+		}
+
+		for _, testDomain := range testDomains {
+			startTime := time.Now()
+			testAddr := net.JoinHostPort(testDomain, "443")
+
+			f.logCtx(ctx, "🏃 run Fallback: '%v' (domain: %v)\n", fallbackUrl, testDomain)
+
+			ctx, cancel := context.WithTimeout(ctx, f.TestTimeout)
+			defer cancel()
+			testConn, err := dialer.DialStream(ctx, testAddr)
+			if err != nil {
+				f.logCtx(ctx, "🏁 got Fallback: '%v' (domain: %v), duration=%v, dial_error=%v ❌\n", fallbackUrl, testDomain, time.Since(startTime), err)
+			}
+			tlsConn := tls.Client(testConn, &tls.Config{ServerName: testDomain})
+			err = tlsConn.HandshakeContext(ctx)
+			tlsConn.Close()
+			if err != nil {
+				f.logCtx(ctx, "🏁 got Fallback: '%v' (domain: %v), duration=%v, handshake=%v ❌\n", fallbackUrl, testDomain, time.Since(startTime), err)
+				return nil, err
+			} else {
+				f.logCtx(ctx, "🏁 got Fallback: '%v' (domain: %v), duration=%v, status=ok ✅\n", fallbackUrl, testDomain, time.Since(startTime))
+			}
+		}
+		return &SearchResult{dialer, fallbackUrl}, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not find a working fallback: %w", err)
 	}
+	f.log("🏆 selected fallback '%v' in %0.2fs\n\n", fallback.Config, time.Since(raceStart).Seconds())
+	return fallback.Dialer, nil
+}
 
-	resolver, err := f.findDNS(ctx, testDomains, parsedConfig.DNS)
+// Attempts to create a new Dialer using only proxyless (DNS and TLS) strategies
+func (f *StrategyFinder) NewProxylessDialer(ctx context.Context, testDomains []string, config configConfig) (transport.StreamDialer, error) {
```

Comment:
This is not needed. You can just pass a config without proxies. Let's remove to avoid bloat in the API.

Reply:
I had split this out from NewDialer because that function was getting too long. Made it private (lowercase)